zpeclient.go

```diff
@@ -0,0 +1,230 @@
+		if policy == nil || policy.Name == "" {
+			log.Println("policy in Contents.Policies is nil")
+			continue
+		}
+		for _, assertion := range policy.Assertions {
+			if assertion == nil || assertion.Role == "" || assertion.Resource == "" || assertion.Action == "" {
+				log.Println("assertion in policy.Assertions is nil")
+				continue
+			}
+			_, ok := crMap.roleToAssertion[assertion.Role]
+			if !ok {
+				crMap.roleToAssertion[assertion.Role] = []*simpleAssertion{}
+			}
+			effect := assertion.Effect
+			resourceRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Resource)) + "$")
+			if err != nil {
+				log.Printf("Error occurred when converting assertion resource into regex format. Error: %v", err)
+				continue
+			}
+			actionRegex, err := regexp.Compile("^" + replacer.Replace(strings.ToLower(assertion.Action)) + "$")
+			if err != nil {
+				log.Printf("Error occurred when converting assertion action into regex format. Error: %v", err)
+				continue
+			}
+			simpleAssert := simpleAssertion{
+				resource: resourceRegex,
+				action:   actionRegex,
+				effect:   effect,
+			}
+			crMap.roleToAssertion[assertion.Role] = append(crMap.roleToAssertion[assertion.Role], &simpleAssert)
+		}
+	}
+	return nil
+}
+
+// addupdateObj - add and update cr object in cache
+func (c *Cache) addOrUpdateObj(item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := c.domainMap[domainName]
+	if ok {
+		delete(c.domainMap, domainName)
+	}
+	roleToPrincipals := make(map[string][]*simplePrincipal)
+	roleToAssertion := make(map[string][]*simpleAssertion)
+	crMap := roleMappings{
+		roleToPrincipals: roleToPrincipals,
+		roleToAssertion:  roleToAssertion,
+	}
+	c.domainMap[domainName] = crMap
+	err := parseData(c.domainMap, domainName, item, c.log)
+	if err != nil {
+		c.log.Printf("Error happened parsing AthenzDomains CR info. Error: %v", err)
+	}
+	c.lock.Unlock()
+}
+
+// deleteObj - delete object in cache
+func (c *Cache) deleteObj(item *v1.AthenzDomain) {
+	c.lock.Lock()
+	domainName := item.ObjectMeta.Name
+	_, ok := c.domainMap[domainName]
+	if ok {
+		delete(c.domainMap, domainName)
+	}
+	c.lock.Unlock()
+}
+
+// authorize - authorize using cache data
+func (c *Cache) authorize(principal string, check AthenzAccessCheck) (bool, error) {
+	domainName := strings.Split(check.Resource, ":")
+	if len(domainName) != 2 {
+		return false, errors.New("Error splitting domain name")
+	}
+
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	crMap := c.domainMap
+	domainData, ok := crMap[domainName[0]]
+	if !ok {
+		return false, fmt.Errorf("%s does not exist in cache map", domainName[0])
+	}
+	roles := []string{}
+	for role, members := range domainData.roleToPrincipals {
+		for _, member := range members {
+			if member.memberRegex.MatchString(principal) {
+				if member.expiration.IsZero() || member.expiration.After(time.Now()) {
+					roles = append(roles, role)
+				}
+			}
+		}
+	}
+
```

Comment:
could we add a log message to print how many roles matched a given principal, would be helpful in debugging
```
log.Printf("%d roles matched principal: %s on domain: %s", len(roles), principal, domainName[0])
```